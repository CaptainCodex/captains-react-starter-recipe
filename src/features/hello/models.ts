/* Hello */
export interface IHello {
  greeting: string;
}

export class Hello implements IHello {
  constructor(public greeting: string) {}

  static create(json: Hello) {
    return new Hello(json.greeting);
  }
}

/* World */
export interface IWorld {
  salutation: string;
}

export class World implements IWorld {
  constructor(public salutation: string) {}

  static create(json: World) {
    return new World(json.salutation);
  }
}
