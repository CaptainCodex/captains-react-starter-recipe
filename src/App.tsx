import * as React from 'react';
import { Component } from 'react';
import { BrowserRouter, Switch, Route } from 'react-router-dom';
import { ReactCSSTransitionGroup } from 'react-addons-css-transition-group';
import { Home } from './pages/Home';

class App extends Component {
  render() {
    return (
      <BrowserRouter>
        <Switch>
          <Route exact path="/" component={Home} />
        </Switch>
      </BrowserRouter>
    );
  }
}

export default App;
