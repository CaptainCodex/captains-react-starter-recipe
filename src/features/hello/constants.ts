export const GET_HELLO = 'hello/GET_HELLO';
export const GET_HELLO_RESPONSE = 'hello/GET_HELLO_RESPONSE';

export const GET_WORLD_REQUEST = 'hello/GET_WORLD_REQUEST';
export const GET_WORLD_RESPONSE = 'hello/GET_WORLD_RESPONSE';
export const GET_WORLD_ERROR = 'hello/GET_WORLD_ERROR';
