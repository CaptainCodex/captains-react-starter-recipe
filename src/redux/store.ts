import { createStore, applyMiddleware } from 'redux';
import { createEpicMiddleware } from 'redux-observable';
import logger from 'redux-logger';

import Types from 'Types';
import { composeEnhancers } from './utils';
import rootReducer from './root-reducer';
import rootEpic from './root-epic';
import services from '../services';

export const epicMiddleware = createEpicMiddleware<
  Types.RootAction,
  Types.RootAction,
  Types.RootState
>({
  dependencies: services
});

function configureStore(initialState?: object) {
  const middlewares = [epicMiddleware, logger];

  if (process.env.NODE_ENV !== 'development') {
    middlewares.pop();
  }

  const enhancer = composeEnhancers(applyMiddleware(...middlewares));

  return createStore(rootReducer, initialState, enhancer);
}

const store = configureStore();

epicMiddleware.run(rootEpic);

export default store;
