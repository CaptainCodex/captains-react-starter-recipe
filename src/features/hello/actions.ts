import { createAction, createAsyncAction } from 'typesafe-actions';

import * as contants from './constants';
import { AjaxResponse, AjaxError } from 'rxjs/ajax';

export const getHello = createAction(contants.GET_HELLO);
export const getHelloResponse = createAction(
  contants.GET_HELLO_RESPONSE,
  resolve => (response: AjaxResponse) => resolve(response)
);

export const getWorld = createAsyncAction(
  contants.GET_WORLD_REQUEST,
  contants.GET_WORLD_RESPONSE,
  contants.GET_WORLD_ERROR
)<undefined, AjaxResponse, AjaxError>();
