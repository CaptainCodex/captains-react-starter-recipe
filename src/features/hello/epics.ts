import Types from 'Types';
import { Epic } from 'redux-observable';
import { filter, switchMap } from 'rxjs/operators';
import { isActionOf } from 'typesafe-actions';

import { getHello, getHelloResponse, getWorld } from './actions';

/* 
  Note: Use 'switchMap' to cancel previous ajax calls.
        Use 'mergeMap' to launch multiple concurrent connections.
*/

const getHelloEpic: Epic<
  Types.RootAction,
  Types.RootAction,
  Types.RootState,
  Types.Services
> = (action$, store, { ajaxObservable }) =>
  action$.pipe(
    filter(isActionOf(getHello)),
    switchMap(action =>
      ajaxObservable({
        request: { url: `/hello` },
        retryConfig: {
          initialInterval: 1000,
          maxRetries: 0
        },
        onResponseAction: getHelloResponse
      })
    )
  );

const getWorldEpic: Epic<
  Types.RootAction,
  Types.RootAction,
  Types.RootState,
  Types.Services
> = (action$, store, { ajaxObservable }) =>
  action$.pipe(
    filter(isActionOf(getWorld.request)),
    switchMap(action =>
      ajaxObservable({
        request: { url: `/world` },
        retryConfig: {
          initialInterval: 1000,
          maxRetries: 0
        },
        onResponseAction: getWorld.success,
        onErrorAction: getWorld.failure
      })
    )
  );

export default [getHelloEpic, getWorldEpic];
