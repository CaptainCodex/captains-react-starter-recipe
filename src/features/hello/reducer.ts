import { combineReducers } from 'redux';
import { ActionType } from 'typesafe-actions';

import * as actions from './actions';
import * as constants from './constants';
import { Hello, World } from './models';

export type Action = ActionType<typeof actions>;

export type State = {
  readonly hello: Hello;
  readonly world: World;
};

const initialHello = Hello.create({ greeting: 'Welcome to the app.' });
const initialSalutation = World.create({ salutation: 'Bonjour.' });

export default combineReducers<State, Action>({
  hello: (state = initialHello, action) => {
    switch (action.type) {
      case constants.GET_HELLO_RESPONSE:
        if (state.greeting === 'Welcome to the app.') {
          return new Hello('Change anything you need!');
        } else if (state.greeting === 'Change anything you need!') {
          return new Hello('And happy coding.');
        }

      default:
        return state;
    }
  },
  world: (state = initialSalutation, action) => {
    switch (action.type) {
      case constants.GET_WORLD_RESPONSE:
        return new World('Aloha!');
      case constants.GET_WORLD_ERROR:
        return new World('Error raised.');

      default:
        return state;
    }
  }
});
