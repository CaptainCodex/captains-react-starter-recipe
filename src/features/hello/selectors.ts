import { createSelector } from 'reselect';
import _ from 'lodash';
import { State } from './reducer';

const getGreeting = (state: State) => state.hello.greeting;

export const getCapitalizedGreeting = createSelector(
  [getGreeting],
  greeting => {
    return greeting.toUpperCase();
  }
);
