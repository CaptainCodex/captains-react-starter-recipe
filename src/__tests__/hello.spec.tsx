import * as React from 'react';
import * as nock from 'nock';
import { mount, configure } from 'enzyme';
import { Provider } from 'react-redux';
import App from '../App';
import reducer, { State } from '../features/hello/reducer';
import * as actions from '../features/hello/actions';
import store from '../redux';
import { defer } from '../../utils';

import * as Adapter from 'enzyme-adapter-react-16';
import { getCapitalizedGreeting } from '../features/hello/selectors';

configure({ adapter: new Adapter() });

const render = () =>
  mount(
    <Provider store={store}>
      <App />
    </Provider>
  );

const getInitialState = (initial?: Partial<State>) =>
  reducer(initial as State, {} as any);

describe('Hello reducer', () => {
  describe('initial state', () => {
    it('should match previously saved snapshot', () => {
      const initialState = getInitialState();

      expect(initialState).toMatchSnapshot();
    });
  });

  describe('after calling getHello', () => {
    it('greeting should not change on error', async () => {
      const originalGreeting = store.getState().hello.hello.greeting;

      nock('http://localhost:8080')
        .get('/hello')
        .replyWithError('Something went wrong.');

      store.dispatch(actions.getHello());

      await defer();

      expect(originalGreeting).toEqual(store.getState().hello.hello.greeting);
    });

    it('greeting should change', async () => {
      const originalGreeting = store.getState().hello.hello.greeting;

      nock('http://localhost:8080')
        .get('/hello')
        .reply(200);

      store.dispatch(actions.getHello());

      await defer();

      expect(originalGreeting).not.toEqual(
        store.getState().hello.hello.greeting
      );
    });
  });

  describe('after calling getWorld', () => {
    it('salutation should change', async () => {
      const originalGreeting = store.getState().hello.hello.greeting;

      nock('http://localhost:8080')
        .get('/world')
        .reply(200);

      store.dispatch(actions.getWorld.request());

      await defer();

      expect(originalGreeting).not.toEqual(
        store.getState().hello.world.salutation
      );
    });

    it('salutation should raise an error', async () => {
      nock('http://localhost:8080')
        .get('/world')
        .replyWithError('Something went wrong.');

      store.dispatch(actions.getWorld.request());

      await defer();

      expect(store.getState().hello.world.salutation).toEqual('Error raised.');
    });
  });

  describe('Hello World page', () => {
    it('greeting should be welcoming', () => {
      const wrapper = render();

      const greeting = wrapper.find('#getHello').text();

      expect(greeting).toContain(
        getCapitalizedGreeting(store.getState().hello)
      );
    });

    it('greeting should change after click', async () => {
      const originalGreeting = getCapitalizedGreeting(store.getState().hello);

      nock('http://localhost:8080')
        .get(`/hello`)
        .reply(200);

      const wrapper = render();

      const clickable = wrapper.find('#getHello');

      clickable.simulate('click');

      await defer();

      wrapper.update();

      const greeting = wrapper.find('#getHello').text();

      expect(greeting).not.toEqual(originalGreeting);
    });

    it('salutation should change after click', async () => {
      const originalSalutation = store.getState().hello.world.salutation;

      nock('http://localhost:8080')
        .get(`/world`)
        .reply(200);

      const wrapper = render();

      const clickable = wrapper.find('#getWorld');

      clickable.simulate('click');

      await defer();

      wrapper.update();

      const greeting = wrapper.find('#getWorld').text();

      expect(greeting).not.toEqual(originalSalutation);
    });
  });
});
