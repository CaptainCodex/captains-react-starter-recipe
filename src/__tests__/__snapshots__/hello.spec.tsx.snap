// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hello reducer initial state should match previously saved snapshot 1`] = `
Object {
  "hello": Hello {
    "greeting": "Welcome to the app.",
  },
  "world": World {
    "salutation": "Bonjour.",
  },
}
`;
